server/auth/sso/sso.go

```diff
@@ -0,0 +1,169 @@
+	return &sso{config, idTokenVerifier, baseHRef, secure}, nil
+}
+
+const stateCookieName = "oauthState"
+
+func (s *sso) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	state := rand.RandString(10)
+	http.SetCookie(w, &http.Cookie{
+		Name:     stateCookieName,
+		Value:    state,
+		Expires:  time.Now().Add(1 * time.Hour),
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+		Secure:   s.secure,
+	})
+	http.Redirect(w, r, s.config.AuthCodeURL(state), http.StatusFound)
+}
+
+func (s *sso) HandleCallback(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	state := r.URL.Query().Get("state")
+	cookie, err := r.Cookie(stateCookieName)
+	http.SetCookie(w, &http.Cookie{Name: stateCookieName, MaxAge: 0})
+	if err != nil {
+		w.WriteHeader(400)
+		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %v", err)))
+		return
+	}
+	if state != cookie.Value {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %s", state)))
+		return
+	}
+	oauth2Token, err := s.config.Exchange(ctx, r.URL.Query().Get("code"))
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to exchange token: %v", err)))
+		return
+	}
+	rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to get id_token")))
+		return
+	}
+	idToken, err := s.idTokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to verify token: %v", err)))
+		return
+	}
+	c := &claims{}
+	if err := idToken.Claims(c); err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to get claims: %v", err)))
+		return
+	}
+	token, err := zjwt.ZJWT(rawIDToken)
+	if err != nil {
+		w.WriteHeader(500)
+		_, _ = w.Write([]byte(fmt.Sprintf("failed to get compress token: %v", err)))
+		return
+	}
+	value := Prefix + token
+	log.Debugf("handing oauth2 callback %v", value)
+	http.SetCookie(w, &http.Cookie{
+		Value:    value,
+		Name:     "authorization",
+		Path:     s.baseHRef,
+		Expires:  time.Now().Add(10 * time.Hour),
+		SameSite: http.SameSiteStrictMode,
+		Secure:   s.secure,
+	})
+	http.Redirect(w, r, s.baseHRef, 302)
+}
+
+// authorize verifies a bearer token and pulls user information form the claims.
+func (s *sso) Authorize(ctx context.Context, authorisation string) (wfv1.User, error) {
+	rawIDToken, err := zjwt.JWT(strings.TrimPrefix(authorisation, Prefix))
+	if err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to decompress token %v", err)
+	}
+	idToken, err := s.idTokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to verify id_token %v", err)
+	}
+	c := &claims{}
+	if err := idToken.Claims(c); err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to parse claims: %v", err)
+	}
+	return wfv1.User{Name: idToken.Subject, Groups: c.Groups}, nil
```

Comment:
I don't think we should have an API to return an invented `User` object. `groups` is not a required OIDC field and will not always be returned. Mapping `sub` to `name` is also not any standard. In our experience, there are many users who do not use the `groups` field and key off of a different or custom field for RBAC. For this to be future proof, we really we should return back the raw claims.